parquet-column/src/main/java/parquet/schema/Types.java

```diff
@@ -555,6 +622,528 @@ protected GroupType build(String name) {
+    public static class MapGroupKeyBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseGroupBuilder<P, MapGroupKeyBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapGroupKeyBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+        repetition(Type.Repetition.REQUIRED);
+      }
+
+      @Override
+      protected MapGroupKeyBuilder<P, Q> self() {
+        return this;
+      }
+
+      public MapValueBuilder<P, Q> value(PrimitiveTypeName type, Type.Repetition repetition) {
+        parent.setKeyType(build("key"));
+        if (parent.parent != null) {
+          return new MapValueBuilder<P, Q>(parent,
+              type).repetition(repetition);
+        } else {
+          return new MapValueBuilder<P, Q>(parent,
+              parent.returnType, type).repetition(repetition);
+        }
+      }
+
+      public MapValueBuilder<P, Q> requiredValue(PrimitiveTypeName type) {
+        return value(type, Type.Repetition.REQUIRED);
+      }
+
+      public MapValueBuilder<P, Q> optionalValue(PrimitiveTypeName type) {
+        return value(type, Type.Repetition.OPTIONAL);
+      }
+
+      public MapGroupValueBuilder<P, Q> value(Type.Repetition repetition) {
+        parent.setKeyType(build("key"));
+        return new MapGroupValueBuilder<P, Q>(parent).repetition(repetition);
+      }
+
+      public MapGroupValueBuilder<P, Q> requiredGroupValue() {
+        return value(Type.Repetition.REQUIRED);
+      }
+
+      public MapGroupValueBuilder<P, Q> optionalGroupValue() {
+        return value(Type.Repetition.OPTIONAL);
+      }
+    }
+
+    public static class MapGroupValueBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseGroupBuilder<P, MapGroupValueBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapGroupValueBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapGroupValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    protected void setKeyType(Type keyType) {
+      this.keyType = keyType;
+    }
+
+    protected void setValueType(Type valueType) {
+      this.valueType = valueType;
+    }
+
+    private Type keyType = null;
+    private Type valueType = null;
+    private Class<P> returnType = null;
+
+    public BaseMapBuilder(P parent) {
+      super(parent);
+    }
+
+    private BaseMapBuilder(Class<P> returnType) {
+      super(returnType);
+      this.returnType = returnType;
+    }
+
+    @Override
+    protected abstract Q self();
+
+    public MapKeyBuilder<P, Q> key(PrimitiveTypeName type) {
+      if (parent != null) {
+        return new MapKeyBuilder<P, Q>(this, type);
+      } else {
+        return new MapKeyBuilder<P, Q>(this, returnType, type);
```

Comment:
value() inside MapKeyBuilder uses MapBuilder's parent. So, it is required
:). I should have mentioned in the first reply itself.

On Mon, Mar 23, 2015 at 5:16 PM, Ryan Blue <notifications@github.com> wrote:

> In parquet-column/src/main/java/parquet/schema/Types.java
> <https://github.com/apache/incubator-parquet-mr/pull/148#discussion_r26994189>
> :
>
> > +      super(parent);
> > +    }
> > +
> > +    private BaseMapBuilder(Class<P> returnType) {
> > +      super(returnType);
> > +      this.returnType = returnType;
> > +    }
> > +
> > +    @Override
> > +    protected abstract Q self();
> > +
> > +    public MapKeyBuilder<P, Q> key(PrimitiveTypeName type) {
> > +      if (parent != null) {
> > +        return new MapKeyBuilder<P, Q>(this, type);
> > +      } else {
> > +        return new MapKeyBuilder<P, Q>(this, returnType, type);
>
> I don't think this is needed since the Map builder's parent is no longer
> explicitly used. The other builders don't have it either.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-parquet-mr/pull/148/files#r26994189>.
>



-- 

Regards,
Ashish